### ISSUE TITLE
Visualizer: Error when trying to solve from visualizer

### ISSUE THREAD

**Issue by davidabek1** [potassco/asprilo#31]
Debian 9(stretch), installed asprilo via conda
visualizer 0.2.0
clingo 5.3.0
'async_' is wrong key argument to clingo.control.solve(on_model = self.on_model, async_ = True)
 in solver.py (egg file)

the error shows:
solve: 0
Traceback (most recent call last):
  File "/home/david/anaconda3/envs/asprilo/bin/viz-solver", line 10, in <module>
    sys.exit(main())
  File "/home/david/anaconda3/envs/asprilo/lib/python3.6/site-packages/visualizer-0.2.2-py3.6.egg/visualizer/solver.py", line 366, in main
  File "/home/david/anaconda3/envs/asprilo/lib/python3.6/site-packages/visualizer-0.2.2-py3.6.egg/visualizer/solver.py", line 250, in run
  File "/home/david/anaconda3/envs/asprilo/lib/python3.6/site-packages/visualizer-0.2.2-py3.6.egg/visualizer/solver.py", line 138, in receive
  File "/home/david/anaconda3/envs/asprilo/lib/python3.6/site-packages/visualizer-0.2.2-py3.6.egg/visualizer/solver.py", line 164, in on_raw_data
  File "/home/david/anaconda3/envs/asprilo/lib/python3.6/site-packages/visualizer-0.2.2-py3.6.egg/visualizer/solver.py", line 201, in on_data
  File "/home/david/anaconda3/envs/asprilo/lib/python3.6/site-packages/visualizer-0.2.2-py3.6.egg/visualizer/solver.py", line 277, in solve
TypeError: 'async_' is an invalid keyword argument for this function
close solver

**Comment by pobermei** [comment]
This happens since the visualizer requires clingo's current development version 5.4.0 where keyword `async` of `clingo.Control.solve` was changed to `async_` for compatibility with python>=3.7. For proper installation, call `conda install asprilo-visualizer -c potassco -c potassco/label/dev` to also pull in clingo 5.4.0. You can also find a detailed install guide [here](https://asprilo.github.io/visualizer/).

**Comment by davidabek1** [comment]
Thank you for answering and updating the installation instructions, 
after remove and re-install clingo 5.4.0, now i get to point where i get message "found solution", 
and the plan is not loaded into the visualizer.
should i open a new issue?

here a snippet from my terminal:
Connection with: ('127.0.0.1', 45230)
/home/david/asprilo/encodings/m/./action-M.ilp:19:43-54: info: atom does not occur in any rule head:
  move(#X0,#P1,#X2)

/home/david/asprilo/encodings/m/./goal-M.ilp:7:22-38: info: atom does not occur in any rule head:
  processed(#X0,#P1,#X2)

found solution

**Comment by pobermei** [comment]
Can you please provide the instance data you are using?
Maybe @Thomas-Otto can comment on this too.

Alternatively, I'd recommend to not use a solver script at first but simply load or pipe your instance and computed answer set into the visualizer.

**Comment by davidabek1** [comment]
Thank you, I've already done piping as a workaround and it works,
But since we (students team) aim on extending Asprilo with more solvers we want it to be embedded in the visualizer.

BR,
David.

________________________________

**Comment by davidabek1** [comment]
@Thomas-Otto and @pobermei , i've attached the information on instance data:
settings solver instantiate: viz-solver --port 5000 -m incremental -e ~/asprilo/encodings/m/encoding.ilp
(encodings have been downloaded from your git)
I've changed the file suffix from lp to txt so it uploads directly and not a zip file.

[demo_mapf_solver_withProd_Orders_WithPickingStations.txt](https://github.com/potassco/asprilo/files/3348401/demo_mapf_solver_withProd_Orders_WithPickingStations.txt)

**Comment by Thomas-Otto** [comment]
Hey. It took me some time to figure that out. The problem seems to be an incompatibility between the 'inc' mode and the solver. The solver works slightly different and the encodings are build for the 'inc' mode from clingo. But adding the following two lines to the goal-M.ilp file under the "#program check(t)" section should fix the problem:
#external query(t).
occurs(object(robot,R), action(move,D),     t) :- move(robot(R),D,t).
You can add these two lines to every other program part that iterates through the time steps, too.

**Comment by davidabek1** [comment]
It works!!
thank you,
@Thomas-Otto as you directed I've changed goal-M.ilp, and it was enough to get it to work, 
but you mentioned that i can add this 2 lines to every other program part that iterates time steps, 
besides goal-M.ilp and goal-M-XY.ilp what are the other files to be candidate for this change and when it will show (as currently it works fine, with change only to goal-M.ilp)?

**Comment by davidabek1** [comment]
Hi @Thomas-Otto , continued my question from previous comment, I **again tackle kind of the same issue**.
We have added the ability to assign robots to specific shelves, in order to transform the problems from anonymous to non-anonymous (based on your team instructions).
Now again Clingo solving the problem and I can pipe it to viz, but doing it from the visualizer I get infinite calls UNSAT, not solving the problem.
Attached the same problem but with added ASSIGN rules, attached all changed encoding files to support it: input.lp, encoding.ilp, goal-M-assign.ilp
[encodings_folder.zip](https://github.com/potassco/asprilo/files/3371382/encodings_folder.zip)
While attaching the zip file, I have tried to change output-M-assign.ilp, with added lines to show assign rule in the output, it made it worse: as i could only solve with Clingo, couldn't pipe to viz (empty everything, as it didn't pass the plan to the pipe), starting from viz show the same issue as infinite calls UNSAT.

**Comment by Thomas-Otto** [comment]
Hi @davidabek1,
i meant that it does not matter to which file you add these lines to. It is just a matter of style and how you want to structure your code. It should also work if you add the lines to the "action-M.ilp" file.

The way the visualizer works is:
When it receives an instance it reads this instances in asp atoms and translates it into an internal model. When is sends an instance to the solver it translates this internal model back into asp atoms and send these atoms to the solver. So the visualizer does not simply sends whole instance files to the solver. Atoms that the visualizer can not handle are discarded.
The visualizer expects that every input atom follows a specific syntax:
init(object([OBJECTNAME], [OBJECTID]), value([VALUENAME], [VALUEVALUE])).
occurs(object([OBJECTNAME], [OBJECTID]), action([ACTIONNAME], [ACTIONVALUE])).
However every atom that the visualizer can't handle but matches this syntax will be remembered by the visualizer and will be send to the solver, too.
Long story short:
Your assign atoms in the instance file are ignored by the visualizer and are not send to the solver.
You must change the syntax of your atoms to something like:
init(object(robot, RID), value(assign, (SID, STATIONID))) or
init(object(assign, AID), value(assign, (RID, SID, STATIONID))).
Or you just add something like
"init(object(robot, RID), value(assign, (SID, STATIONID))) :- assign(robot(RID), shelf(SID), station(STATIONID))." 
to your instance file and something like
"assign(robot(RID), shelf(SID), station(STATIONID)) :- init(object(robot, RID), value(assign, (SID, STATIONID)))." 
to the "input-lp" file.
